xds/internal/balancer/gracefulswitch/gracefulswitch_test.go

```diff
@@ -0,0 +1,1125 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package gracefulswitch
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/internal/grpcsync"
+	"google.golang.org/grpc/internal/grpctest"
+	"google.golang.org/grpc/internal/testutils"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+const (
+	defaultTestTimeout      = 5 * time.Second
+	defaultTestShortTimeout = 10 * time.Millisecond
+)
+
+type s struct {
+	grpctest.Tester
+}
+
+func Test(t *testing.T) {
+	grpctest.RunSubTests(t, s{})
+}
+
+func setup(t *testing.T) (*testutils.TestClientConn, *Balancer) {
+	tcc := testutils.NewTestClientConn(t)
+	return tcc, NewBalancer(tcc, balancer.BuildOptions{})
+}
+
+// TestSuccessfulFirstUpdate tests a basic scenario for the Graceful Switch Load
+// Balancer, where it is setup with a balancer which should populate the Current
+// Load Balancer. Any ClientConn updates should then be forwarded to this
+// Current Load Balancer.
+func (s) TestSuccessfulFirstUpdate(t *testing.T) {
+	_, gsb := setup(t)
+	if err := gsb.SwitchTo(mockBalancerBuilder1{}); err != nil {
+		t.Fatalf("Balancer.SwitchTo failed with error: %v", err)
+	}
+	if gsb.balancerCurrent == nil {
+		t.Fatal("balancerCurrent was not built out when a correct SwitchTo() call should've triggered the balancer to build")
+	}
+	// This will be used to update the Graceful Switch Balancer. This update
+	// should simply be forwarded down to the Current Load Balancing Policy.
+	ccs := balancer.ClientConnState{
+		BalancerConfig: mockBalancerConfig{},
+	}
+
+	// Updating ClientConnState should forward the update exactly as is to the
+	// current balancer.
+	if err := gsb.UpdateClientConnState(ccs); err != nil {
+		t.Fatalf("Balancer.UpdateClientConnState(%v) failed with error: %v", ccs, err)
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := gsb.balancerCurrent.Balancer.(*mockBalancer).waitForClientConnUpdate(ctx, ccs); err != nil {
+		t.Fatalf("error in ClientConnState update: %v", err)
+	}
+}
+
+// TestTwoBalancersSameType tests the scenario where there is a graceful switch
+// load balancer setup with a current and pending Load Balancer of the same
+// type. Any ClientConn update should be forwarded to the current if no pending,
+// and the only the pending if the graceful switch balancer contains both a
+// current and pending. The pending Load Balancer should also swap into current
```

Comment:
> // Any ClientConn update should be forwarded to the current if no pending,
// and the only the pending if the graceful switch balancer contains both a
// current and pending

This comment needs some fixing.

Reply:
Switched to "Any ClientConn update should be forwarded to the current lb if there is a current lb and no pending lb, and the only the pending lb if the graceful switch balancer contains both a current lb and a pending lb."